src/Analyzers/CSharp/Analyzers/UseSystemThreadingLock/CSharpUseSystemThreadingLockDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,235 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseSystemThreadingLock;
+
+/// <summary>
+/// Looks for code of the form:
+/// 
+///     private ... object _gate = new object();
+///     ...
+///     lock (_gate)
+///     {
+///     }
+///     
+/// and converts it to:
+/// 
+///     private ... Lock _gate = new Lock();
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed class CSharpUseSystemThreadingLockDiagnosticAnalyzer()
+    : AbstractBuiltInCodeStyleDiagnosticAnalyzer(
+        IDEDiagnosticIds.UseSystemThreadingLockDiagnosticId,
+        EnforceOnBuildValues.UseSystemThreadingLock,
+        CSharpCodeStyleOptions.PreferSystemThreadingLock,
+        new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Use_System_Threading_Lock), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+{
+    /// <summary>
+    /// A method body edit anywhere in a type will force us to reanalyze the whole type.
+    /// </summary>
+    /// <returns></returns>
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(compilationContext =>
+        {
+            var compilation = compilationContext.Compilation;
+
+            // The new 'Lock' feature is only supported in C# 13 and above, and only if we actually have a definition of
+            // System.Threading.Lock available.
+            if (!compilation.LanguageVersion().IsCSharp13OrAbove())
+                return;
+
+            var lockType = compilation.GetTypeByMetadataName("System.Threading.Lock");
+            if (lockType is null)
+                return;
+
+            context.RegisterSymbolStartAction(AnalyzeNamedType, SymbolKind.NamedType);
+        });
+    }
+
+    private void AnalyzeNamedType(SymbolStartAnalysisContext context)
+    {
+        var cancellationToken = context.CancellationToken;
+
+        var namedType = (INamedTypeSymbol)context.Symbol;
+        if (namedType is not
+            {
+                TypeKind: TypeKind.Class or TypeKind.Struct,
+                DeclaringSyntaxReferences: [var reference, ..]
```

Comment:
This can cause issues when a partial type has some parts that are generated code and other parts which are not. The analysis should apply specifically to the location where the field is declared.

Reply:
sure.  i'll update to that.

Reply:
done.